pkg/virtual/framework/rootapiserver/root_apiserver.go

```diff
@@ -147,67 +144,39 @@ func (readys asHealthCheck) Check(req *http.Request) error {
 	return nil
 }
 
-func (c completedConfig) resolveRootPaths(urlPath string, requestContext context.Context) (accepted bool, prefixToStrip string, completedContext context.Context) {
-	completedContext = requestContext
-	for _, virtualWorkspace := range c.ExtraConfig.VirtualWorkspaces {
-		if accepted, prefixToStrip, completedContext := virtualWorkspace.ResolveRootPath(urlPath, requestContext); accepted {
-			return accepted, prefixToStrip, context.WithValue(completedContext, virtualcontext.VirtualWorkspaceNameKey, virtualWorkspace.GetName())
-		}
-	}
-	return
-}
-
 func (c completedConfig) getRootHandlerChain(delegateAPIServer genericapiserver.DelegationTarget) func(http.Handler, *genericapiserver.Config) http.Handler {
 	return func(apiHandler http.Handler, genericConfig *genericapiserver.Config) http.Handler {
-		return genericapiserver.DefaultBuildHandlerChain(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			requestContext := req.Context()
 			// detect old kubectl plugins and inject warning headers
 			if req.UserAgent() == "Go-http-client/2.0" {
 				// TODO(sttts): in the future compare the plugin version to the server version and warn outside of skew compatibility guarantees.
-				warning.AddWarning(req.Context(), "",
+				warning.AddWarning(requestContext, "",
 					fmt.Sprintf("You are using an old kubectl-kcp plugin. Please update to a version matching the kcp server version %q.", componentbaseversion.Get().GitVersion))
 			}
 
-			if accepted, prefixToStrip, context := c.resolveRootPaths(req.URL.Path, req.Context()); accepted {
-				req.URL.Path = strings.TrimPrefix(req.URL.Path, prefixToStrip)
-				req.URL.RawPath = strings.TrimPrefix(req.URL.RawPath, prefixToStrip)
-				req = req.WithContext(context)
-				delegatedHandler := delegateAPIServer.UnprotectedHandler()
-				if delegatedHandler != nil {
-					delegatedHandler.ServeHTTP(w, req)
+			for _, virtualWorkspace := range c.ExtraConfig.VirtualWorkspaces {
+				if accepted, prefixToStrip, completedContext := virtualWorkspace.ResolveRootPath(req.URL.Path, requestContext); accepted {
```

Comment:
The only thing that I think maybe worries me a little about this approach is that this work is being done unauthenticated. 

I can't think of a way that this would be a problem right off the bat but just wanted to bring it up.